src/coreclr/jit/lower.cpp

```diff
@@ -5254,13 +5260,9 @@ bool Lowering::LowerUnsignedDivOrMod(GenTreeOp* divMod)
             BlockRange().InsertBefore(divMod, preShiftBy, adjustedDividend);
             firstNode = preShiftBy;
         }
-        else if (type != TYP_I_IMPL
-#ifdef TARGET_ARM64
-                 && !simpleMul // On ARM64 we will use a 32x32->64 bit multiply as that's faster.
-#endif
-                 )
+        else if (widenToNativeIntForMul)
         {
-            adjustedDividend = comp->gtNewCastNode(TYP_I_IMPL, adjustedDividend, true, TYP_U_IMPL);
+            adjustedDividend = comp->gtNewCastNode(TYP_I_IMPL, adjustedDividend, true, TYP_I_IMPL);
```

Comment:
For my education, why is `TYP_I_IMPL` better here than `TYP_U_IMPL`? Other `gtNewCastNode` uses seem to use the unsigned type in such cases.

Reply:
This is a case of redundant information. For `int -> long` casts (non-overflowing) `TYP_LONG` and `TYP_ULONG` are equivalent, only the `fromUnsigned` bit matters and decides whether we'll be sign extending or zero extending.

There is code in the compiler (you can see it in the importer especially) which tries to match up `fromUnsigned` with `TYP_ULONG` - it is a remnant of a bygone era when there was buggy code which used the target type for the `zext`/`sext` decision (also the fact that casts from floating point types _do_ care about the destination type).

Since it is better to converge on only one representation of something in IR, and IR uses signed types for most things, it is better to use `TYP_LONG` (for example, there is code in `optNarrowTree` which misses some folding opportunities because of `TYP_ULONG` in casts, there also was code in assertion propagation that could lead to silent bad codegen for sort of the same reason). I have a branch which will turn using `TYP_ULONG/TYP_UINT` where they are not needed into an assert, but it is yet to PRed.